### PR TITLE
Allow empty obs in non-dominated space partition

### PR DESCRIPTION
**Pull request by TsingQAQ**
This is a PR for allowing empty/None observation specified for non-dominated partition. In some cases, there maybe no observations below a specified reference point. In this scenario, this functionality will just return a hyper-cell (bounds) that has been constructed by anti_reference and reference point, driving greedy acquisition function search within this region. 

**Main Modifications**
- change `observations` to optional arg
- change the order of args of `prepare_default_non_dominated_partition_bounds`
- Add return case when `observations` is not specified (or empty)

This PR is the 1st step of tackling #353 (option 1).

**Note**: we leave the responsibility of screening observations below reference point within `prepare_acquisition_function`.

### CODE REVIEW

File: trieste/acquisition/multi_objective/partition.py [hunk 1]
```diff
@@ -45,35 +49,48 @@ def prepare_default_non_dominated_partition_bounds(
...
+    ):  # if anti_reference point is not specified, use a -1e10 as default (act as -inf)
+        anti_reference = specify_default_anti_reference_point(reference, observations)
+    else:  # anti_reference point is specified
+        tf.debugging.assert_shapes([(anti_reference, ["D"])])
+
+    if not_valid_obs(observations):  # if no valid observations
+        assert tf.reduce_all(tf.less_equal(anti_reference, reference)), ValueError(
+            "anti_reference points contains dimension bigger than reference point"
+        )
+        return tf.expand_dims(anti_reference, 0), tf.expand_dims(reference, 0)
```
Comment (vpicheny) [inline trieste/acquisition/multi_objective/partition.py:93]: is it OK that this doesn't return a `_BoundIndexPartition`?

File: trieste/acquisition/multi_objective/partition.py [hunk 2]
```diff
@@ -45,35 +49,48 @@ def prepare_default_non_dominated_partition_bounds(
...
+    tf.debugging.assert_shapes([(reference, ["D"])])
+    if (  # prepare default anti_reference point if not specified
+        anti_reference is None
+    ):  # if anti_reference point is not specified, use a -1e10 as default (act as -inf)
+        anti_reference = specify_default_anti_reference_point(reference, observations)
+    else:  # anti_reference point is specified
+        tf.debugging.assert_shapes([(anti_reference, ["D"])])
+
+    if not_valid_obs(observations):  # if no valid observations
+        assert tf.reduce_all(tf.less_equal(anti_reference, reference)), ValueError(
```
Comment (vpicheny) [inline trieste/acquisition/multi_objective/partition.py:89]: isn't this something we want to test all the time? (i.e. whether or not observations are valid or not)

File: trieste/acquisition/multi_objective/partition.py [hunk 3]
```diff
@@ -45,35 +49,48 @@ def prepare_default_non_dominated_partition_bounds(
...
         tf.debugging.assert_greater_equal(
-            reference,
-            anti_reference,
+            ref,
+            anti_ref,
             message="reference point containing dimensionality below default "
             "anti-reference point ([-1e10, ..., -1e10]), try specify a lower "
             "anti-reference point.",
         )
+        if not not_valid_obs(obs):  # make sure given (valid) observations are larger than -1e10
```
Comment (vpicheny) [inline trieste/acquisition/multi_objective/partition.py]: `not_valid` seems a bit misleading. Change to `is_empty_obs`?

File: trieste/acquisition/multi_objective/partition.py (hunk 2 above)
Comment (vpicheny) [reply trieste/acquisition/multi_objective/partition.py:89]: also change message to "anti_reference point contains at least one value larger than reference point" ?

**Review by vpicheny — COMMENTED** [review]
This strategy seems like a good default when the reference dominates the observations

File: trieste/acquisition/multi_objective/partition.py (hunk 1 above)
Comment (TsingQAQ) [reply trieste/acquisition/multi_objective/partition.py:93]: Was originally expecting calling `prepare_default_non_dominated_partition_bounds` is the same as calling a `any_partition().partition_bounds()`. The latter is expected to return just a tuple of _static_ `lb_points` and `ub_points`.    
So as long as a _static_ `lb_points` & `ub_points` is acceptable for acq, not return a `_BoundIndexPartition_instance`/`any_partition_instance` is ok?

File: trieste/acquisition/multi_objective/partition.py (hunk 2 above)
Comment (TsingQAQ) [reply trieste/acquisition/multi_objective/partition.py:89]: > isn't this something we want to test all the time? (i.e. whether or not observations are valid or not)

Not sure if got the point (please correct if misinterpreted), thought we only need to test when user specified a reference point (as otherwise the current automatic reference selection would always result valid `observation`).

File: trieste/acquisition/multi_objective/partition.py (hunk 1 above)
Comment (vpicheny) [reply trieste/acquisition/multi_objective/partition.py:93]: my bad, I didn't realise we were only using `partition_bounds()`
This is fine then.

File: trieste/acquisition/multi_objective/partition.py (hunk 2 above)
Comment (vpicheny) [reply trieste/acquisition/multi_objective/partition.py:89]: My point was just that we could test `anti_reference << reference` all the time, not only if `not_valid_obs(observations)`.

Anyway this is not very important.

File: trieste/acquisition/multi_objective/partition.py [hunk 4]
```diff
@@ -45,35 +49,48 @@ def prepare_default_non_dominated_partition_bounds(
...
+        return anti_ref
 
-    if tf.shape(observations)[-1] > 2:
+    tf.debugging.assert_shapes([(reference, ["D"])])
+    if (  # prepare default anti_reference point if not specified
+        anti_reference is None
+    ):  # if anti_reference point is not specified, use a -1e10 as default (act as -inf)
+        anti_reference = specify_default_anti_reference_point(reference, observations)
+    else:  # anti_reference point is specified
+        tf.debugging.assert_shapes([(anti_reference, ["D"])])
```
Comment (apaleyes) [inline trieste/acquisition/multi_objective/partition.py]: I find this formatting a bit hard to follow. How about this?

```suggestion
    if (anti_reference is None):
        # if anti_reference point is not specified, use a -1e10 as default (act as -inf)
        anti_reference = specify_default_anti_reference_point(reference, observations)
    else:
        # anti_reference point is specified
        tf.debugging.assert_shapes([(anti_reference, ["D"])])
```

File: trieste/acquisition/multi_objective/partition.py [hunk 5]
```diff
@@ -45,35 +49,48 @@ def prepare_default_non_dominated_partition_bounds(
...
+
+    if is_empty_obs(observations):  # if no valid observations
+        assert tf.reduce_all(tf.less_equal(anti_reference, reference)), ValueError(
+            "anti_reference point contains at least one value larger than reference point"
```
Comment (apaleyes) [inline trieste/acquisition/multi_objective/partition.py]:
```suggestion
            f"anti_reference point {anti_reference} contains at least one value larger than reference point {reference}"
```

**Review by apaleyes — COMMENTED** [review]
That looks good. Just one minor suggestion. There is quite a bit of validation going on (good thing!), but almost all error messages thrown contain no runtime information, e.g. variable values. Having those values in would tremendously help debugging, so if you could augment some error messages with data that caused them, it would be awesome. One idea is below.

File: trieste/acquisition/multi_objective/partition.py (hunk 1 above)
Comment (TsingQAQ) [reply trieste/acquisition/multi_objective/partition.py:93]: Yes, we only provide bounds here. Not immediately aware of a case we need `partition_instance`, unless we need somehow dynamically change partition bounds on the fly during one iteration (if having this use case please indicate), otherwise maybe we can just provide _static_ `partition_bounds` as a "default" choice.

File: trieste/acquisition/multi_objective/partition.py (hunk 2 above)
Comment (TsingQAQ) [reply trieste/acquisition/multi_objective/partition.py:89]: ah i see, was thinking as `prepare_default_non_dominated_partition_bounds` only add a "new" partition strategy for empty observations, hence we only do `anti_reference << reference` in this new partition. As in other cases, this function will invoke other partition method and its these partition methods responsibility to check if `anti_reference << reference` (i.e., [this test](https://github.com/secondmind-labs/trieste/blob/45f99cab9d0bdc5a174fdf5d742f3f52a20ae9ea/tests/unit/acquisition/multi_objective/test_partition.py#L118-L127) and [this test](https://github.com/secondmind-labs/trieste/blob/45f99cab9d0bdc5a174fdf5d742f3f52a20ae9ea/tests/unit/acquisition/multi_objective/test_partition.py#L218-L232))

File: trieste/acquisition/multi_objective/partition.py (hunk 4 above)
Comment (TsingQAQ) [reply trieste/acquisition/multi_objective/partition.py]: Yep, thanks for the suggestions!

**Review by vpicheny — APPROVED** [review]
